keps/sig-api-machinery/20190802-dynamic-coordinated-storage-version.md

```diff
@@ -0,0 +1,338 @@
+status: provisional
+---
+
+# Coordinated dynamic storage version
+
+## Table of Contents
+
+* [Overview](#overview)
+* [API changes](#api-changes)
+   * [Storage Version API](#storage-version-api)
+* [Changes to API servers](#changes-to-api-servers)
+   * [Curating a list of participating API servers in HA master](#curating-a-list-of-participating-api-servers-in-ha-master)
+   * [Voting for Storage Version](#voting-for-storage-version)
+   * [Voting in action](#voting-in-action)
+   * [The selected storage version is supported by all API servers during upgrades/downgrades](#the-selected-storage-version-is-supported-by-all-api-servers-during-upgradesdowngrades)
+   * [Garbage collection of the StorageVersion objects of removed resources](#garbage-collection-of-the-storageversion-objects-of-removed-resources)
+   * [Using the selected storage version when serializing data for etcd](#using-the-selected-storage-version-when-serializing-data-for-etcd)
+   * [Updating the discovery document](#updating-the-discovery-document)
+   * [CRDs](#crds)
+   * [Aggregated API servers](#aggregated-api-servers)
+* [Backwards Compatibility](#backwards-compatibility)
+* [Risks](#risks)
+* [Graduation Plan](#graduation-plan)
+* [FAQ](#faq)
+* [Future work: solving other coordination problems](#future-work-solving-other-coordination-problems)
+* [Alternatives](#alternatives)
+   * [Exposing if API servers in an HA setup are using the same storage version](#exposing-if-api-servers-in-an-ha-setup-are-using-the-same-storage-version)
+   * [Letting the storage migrator detect if API server instances are in agreement](#letting-the-storage-migrator-detect-if-api-server-instances-are-in-agreement)
+* [References](#references)
+
+## Overview
+
+During the rolling update of an HA master, API server instances *i)* use
+different storage versions for a built-in API resource, and *ii)* show different
+storageVersionHash in the discovery documents. These facts make the storage
+migrator not working properly ([details][]).
+
+[storageVersionHash]:https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L979
+[details]:https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/35-storage-version-hash.md#ha-masters
+
+We propose a mechanism for HA API servers to vote on the storage version to use,
+then show the selected storage version in the discovery document and switch to
+use the selected version atomically.
+
+## API changes
+
+### Storage Version API
+
+We introduce a new API `StorageVersion`, in a new API group
+`kube-apiserver.internal.k8s.io/v1alpha1`. 
+
+```golang
+// Storage version for a specific resource.
+type StorageVersion struct {
+  TypeMeta
+  // The name is the GVR, i.e., <group>.<version>.<resource>.
+  ObjectMeta
+  // The selected storage version. All API server intances encode the resource
+  // objects in this version when committing it to etcd.
+  SelectedVersion string
```

Comment:
If we are not configuring storage (I don't think this is what we want), this field is unnecessary.